{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "65f9bb81-4dab-4c10-b9f2-e6b5c04ce0ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# /opt/conda/envs/aneris-dna/lib/python3.14/\n",
    "import os, glob\n",
    "import zipfile\n",
    "\n",
    "# PEMASequenceRetriever\n",
    "# =====\n",
    "param_gene_sequences = \"ERR12541385,ERR12541386,ERR12541387,ERR7125481\"\n",
    "param_fname_seq_zip = \"sequences_compressed.zip\"\n",
    "\n",
    "conf_fname_lib_ena = \"enaDataGet.py\"\n",
    "\n",
    "# PEMARunner\n",
    "# =====\n",
    "param_pema_input_tsv = \"parameters.tsv\"\n",
    "param_pema_input_seq = \"sequences_compressed.zip\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9aa782d2-7520-40d4-a0d6-41d1b5285a34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Package                 Version\n",
      "----------------------- -----------\n",
      "argon2-cffi             25.1.0\n",
      "argon2-cffi-bindings    25.1.0\n",
      "asttokens               3.0.1\n",
      "Brotli                  1.2.0\n",
      "certifi                 2025.11.12\n",
      "cffi                    2.0.0\n",
      "charset-normalizer      3.4.4\n",
      "comm                    0.2.3\n",
      "debugpy                 1.8.17\n",
      "decorator               5.2.1\n",
      "executing               2.2.1\n",
      "h2                      4.3.0\n",
      "hpack                   4.1.0\n",
      "hyperframe              6.1.0\n",
      "idna                    3.11\n",
      "importlib_metadata      8.7.0\n",
      "ipykernel               6.29.5\n",
      "ipython                 9.8.0\n",
      "ipython_pygments_lexers 1.1.1\n",
      "jedi                    0.19.2\n",
      "jupyter_client          8.6.3\n",
      "jupyter_core            5.9.1\n",
      "matplotlib-inline       0.2.1\n",
      "minio                   7.2.15\n",
      "nest_asyncio            1.6.0\n",
      "packaging               25.0\n",
      "parso                   0.8.5\n",
      "pexpect                 4.9.0\n",
      "pip                     25.0.1\n",
      "platformdirs            4.5.0\n",
      "prompt_toolkit          3.0.52\n",
      "psutil                  7.1.3\n",
      "ptyprocess              0.7.0\n",
      "pure_eval               0.2.3\n",
      "pycparser               2.22\n",
      "pycryptodome            3.23.0\n",
      "Pygments                2.19.2\n",
      "PySocks                 1.7.1\n",
      "python-dateutil         2.9.0.post0\n",
      "python-dotenv           1.2.1\n",
      "pyzmq                   27.1.0\n",
      "requests                2.32.5\n",
      "SecretsProvider         2.0.3\n",
      "six                     1.17.0\n",
      "stack_data              0.6.3\n",
      "tornado                 6.5.2\n",
      "traitlets               5.14.3\n",
      "typing_extensions       4.15.0\n",
      "urllib3                 2.5.0\n",
      "wcwidth                 0.2.14\n",
      "zipp                    3.23.0\n",
      "zstandard               0.25.0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\ndir_tmp\\ndir_lib\\n'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "!pip list\n",
    "# !pip install requests\n",
    "\n",
    "# TODO\n",
    "\"\"\"\n",
    "dir_tmp\n",
    "dir_lib\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "caa2628a-d6a2-4701-a7fb-c15b29e0ef53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking availability of https://www.ebi.ac.uk/ena/browser/api/xml/ERR12541385\n",
      "ERR12541385_1.fastq.gz already exists in local directory, skipping\n",
      "ERR12541385_2.fastq.gz already exists in local directory, skipping\n",
      "Completed\n",
      "Checking availability of https://www.ebi.ac.uk/ena/browser/api/xml/ERR12541386\n",
      "ERR12541386_1.fastq.gz already exists in local directory, skipping\n",
      "ERR12541386_2.fastq.gz already exists in local directory, skipping\n",
      "Completed\n",
      "Checking availability of https://www.ebi.ac.uk/ena/browser/api/xml/ERR12541387\n",
      "ERR12541387_1.fastq.gz already exists in local directory, skipping\n",
      "ERR12541387_2.fastq.gz already exists in local directory, skipping\n",
      "Completed\n",
      "Checking availability of https://www.ebi.ac.uk/ena/browser/api/xml/ERR7125481\n",
      "ERR7125481_1.fastq.gz already exists in local directory, skipping\n",
      "ERR7125481_2.fastq.gz already exists in local directory, skipping\n",
      "Completed\n",
      "Compress gene file /home/jovyan/ToDel/1_ena/ERR12541385/ERR12541385_1.fastq.gz\n",
      "Compress gene file /home/jovyan/ToDel/1_ena/ERR12541385/ERR12541385_2.fastq.gz\n",
      "Compress gene file /home/jovyan/ToDel/1_ena/ERR12541386/ERR12541386_1.fastq.gz\n",
      "Compress gene file /home/jovyan/ToDel/1_ena/ERR12541386/ERR12541386_2.fastq.gz\n",
      "Compress gene file /home/jovyan/ToDel/1_ena/ERR12541387/ERR12541387_1.fastq.gz\n",
      "Compress gene file /home/jovyan/ToDel/1_ena/ERR12541387/ERR12541387_2.fastq.gz\n",
      "Compress gene file /home/jovyan/ToDel/1_ena/ERR7125481/ERR7125481_1.fastq.gz\n",
      "Compress gene file /home/jovyan/ToDel/1_ena/ERR7125481/ERR7125481_2.fastq.gz\n",
      "Generated compressed sequencing files /home/jovyan/ToDel/1_ena/ERR7125481/ERR7125481_2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "# PEMASequenceRetriever\n",
    "# =====\n",
    "\n",
    "# test in terminal\n",
    "# python /home/jovyan/library/enaBrowserTools/python3_1.6/enaDataGet.py -f fastq -d /tmp/data/ena ERR12541385\n",
    "\n",
    "# -----\n",
    "dir_cwd = os.getcwd()\n",
    "\n",
    "# dir_tmp = \"/tmp/data\"\n",
    "dir_tmp = os.path.join(dir_cwd, \"ToDel\")\n",
    "\n",
    "dir_lib = os.path.join(dir_cwd, \"library\")\n",
    "\n",
    "# -----\n",
    "dir_lib_ena = os.path.join(dir_lib, \"enaBrowserTools\", \"python3_1.6\")\n",
    "\n",
    "dir_tmp_ena = os.path.join(dir_tmp, \"1_ena\")\n",
    "dir_tmp_seq = os.path.join(dir_tmp, \"2_seq\")\n",
    "\n",
    "try:\n",
    "    os.makedirs(dir_tmp_ena)\n",
    "except FileExistsError:\n",
    "    # directory already exists\n",
    "    pass\n",
    "\n",
    "try:\n",
    "    os.makedirs(dir_tmp_seq)\n",
    "except FileExistsError:\n",
    "    # directory already exists\n",
    "    pass\n",
    "\n",
    "# -----\n",
    "file_lib_ena = os.path.join(dir_lib_ena, conf_fname_lib_ena)\n",
    "file_seq_zip = os.path.join(dir_tmp_seq, param_fname_seq_zip)\n",
    "\n",
    "# -----\n",
    "genes_list = param_gene_sequences.split(',') if len(param_gene_sequences) else []\n",
    "\n",
    "if (len(genes_list) == 0):\n",
    "    raise RuntimeError(\"At least 1 gene sequence must be provided\")\n",
    "\n",
    "for gene in genes_list:\n",
    "    # Run ENA browser tools\n",
    "    # os.system(f\"python {file_lib_ena} -f fastq -d {dir_tmp_ena} {gene}\")\n",
    "    os.system(f\"python {file_lib_ena} -f fastq -d {dir_tmp_ena} -c {gene}\")\n",
    "    \n",
    "    if len(os.listdir(os.path.join(dir_tmp_ena, gene))) != 2:\n",
    "        raise RuntimeError(f\"Expected 2 files for gene `{gene}`. Found {len(os.listdir(dir_tmp_ena))} @ {dir_tmp_ena}\")\n",
    "\n",
    "file_fastq_list = glob.glob(os.path.join(dir_tmp_ena, \"*\", \"*.fastq.gz\"))\n",
    "\n",
    "with zipfile.ZipFile(file_seq_zip, \"w\") as zipObj:\n",
    "    for file_fastq in file_fastq_list:\n",
    "        zipObj.write(file_fastq, os.path.basename(file_fastq))\n",
    "        print(f\"Compress gene file {file_fastq}\")\n",
    "print(f\"Generated compressed sequencing files {file_fastq}\")\n",
    "\n",
    "if not os.path.isfile(file_seq_zip):\n",
    "    raise FileNotFoundError(str(file_seq_zip) + \" is missing\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94bc6c68-9b7e-4736-a91c-260dba04b77c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PEMARunner\n",
    "# =====\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:aneris-dna]",
   "language": "python",
   "name": "conda-env-aneris-dna-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
