{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "65f9bb81-4dab-4c10-b9f2-e6b5c04ce0ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "import zipfile\n",
    "\n",
    "# /opt/conda/envs/aneris-dna/lib/python3.14/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "9aa782d2-7520-40d4-a0d6-41d1b5285a34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: requests in /opt/conda/envs/aneris-dna/lib/python3.14/site-packages (2.32.5)\n",
      "Requirement already satisfied: charset_normalizer<4,>=2 in /opt/conda/envs/aneris-dna/lib/python3.14/site-packages (from requests) (3.4.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /opt/conda/envs/aneris-dna/lib/python3.14/site-packages (from requests) (3.11)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /opt/conda/envs/aneris-dna/lib/python3.14/site-packages (from requests) (2.5.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/envs/aneris-dna/lib/python3.14/site-packages (from requests) (2025.11.12)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\n\\n'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "!pip install requests\n",
    "\n",
    "# TODO\n",
    "\"\"\"\n",
    "dir_tmp\n",
    "dir_lib\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "caa2628a-d6a2-4701-a7fb-c15b29e0ef53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking availability of https://www.ebi.ac.uk/ena/browser/api/xml/ERR12541385\n",
      "Downloading file with md5 check:ftp.sra.ebi.ac.uk/vol1/fastq/ERR125/085/ERR12541385/ERR12541385_1.fastq.gz\n",
      "Downloading file with md5 check:ftp.sra.ebi.ac.uk/vol1/fastq/ERR125/085/ERR12541385/ERR12541385_2.fastq.gz\n",
      "Completed\n",
      "Checking availability of https://www.ebi.ac.uk/ena/browser/api/xml/ERR12541386\n",
      "Downloading file with md5 check:ftp.sra.ebi.ac.uk/vol1/fastq/ERR125/086/ERR12541386/ERR12541386_1.fastq.gz\n",
      "Downloading file with md5 check:ftp.sra.ebi.ac.uk/vol1/fastq/ERR125/086/ERR12541386/ERR12541386_2.fastq.gz\n",
      "Completed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error with FTP transfer: <urlopen error ftp error: >\n",
      "Error with FTP transfer occurred for file: ERR12541387_2.fastq.gz"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking availability of https://www.ebi.ac.uk/ena/browser/api/xml/ERR12541387\n",
      "Downloading file with md5 check:ftp.sra.ebi.ac.uk/vol1/fastq/ERR125/087/ERR12541387/ERR12541387_1.fastq.gz\n",
      "Downloading file with md5 check:ftp.sra.ebi.ac.uk/vol1/fastq/ERR125/087/ERR12541387/ERR12541387_2.fastq.gz\n",
      "Downloading file with md5 check:ftp.sra.ebi.ac.uk/vol1/fastq/ERR125/087/ERR12541387/ERR12541387_2.fastq.gz\n",
      "Completed\n",
      "Checking availability of https://www.ebi.ac.uk/ena/browser/api/xml/ERR7125481\n",
      "Downloading file with md5 check:ftp.sra.ebi.ac.uk/vol1/fastq/ERR712/001/ERR7125481/ERR7125481_1.fastq.gz\n",
      "Downloading file with md5 check:ftp.sra.ebi.ac.uk/vol1/fastq/ERR712/001/ERR7125481/ERR7125481_2.fastq.gz\n",
      "Completed\n",
      "Compress gene file /home/jovyan/ToDel/1_ena/ERR12541385/ERR12541385_1.fastq.gz\n",
      "Compress gene file /home/jovyan/ToDel/1_ena/ERR12541385/ERR12541385_2.fastq.gz\n",
      "Compress gene file /home/jovyan/ToDel/1_ena/ERR12541386/ERR12541386_1.fastq.gz\n",
      "Compress gene file /home/jovyan/ToDel/1_ena/ERR12541386/ERR12541386_2.fastq.gz\n",
      "Compress gene file /home/jovyan/ToDel/1_ena/ERR12541387/ERR12541387_1.fastq.gz\n",
      "Compress gene file /home/jovyan/ToDel/1_ena/ERR12541387/ERR12541387_2.fastq.gz\n",
      "Compress gene file /home/jovyan/ToDel/1_ena/ERR7125481/ERR7125481_1.fastq.gz\n",
      "Compress gene file /home/jovyan/ToDel/1_ena/ERR7125481/ERR7125481_2.fastq.gz\n",
      "Generated compressed sequencing files /home/jovyan/ToDel/1_ena/ERR7125481/ERR7125481_2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "# \n",
    "\n",
    "# test in terminal\n",
    "# python /home/jovyan/library/enaBrowserTools/python3_1.6/enaDataGet.py -f fastq -d /tmp/data/ena ERR12541385\n",
    "\n",
    "# -----\n",
    "param_gene_sequences = \"ERR12541385,ERR12541386,ERR12541387,ERR7125481\"\n",
    "param_fname_seq_zip = \"sequences_compressed.zip\"\n",
    "\n",
    "# -----\n",
    "conf_fname_lib_ena = \"enaDataGet.py\"\n",
    "\n",
    "# -----\n",
    "dir_cwd = os.getcwd()\n",
    "\n",
    "# dir_tmp = \"/tmp/data\"\n",
    "dir_tmp = os.path.join(dir_cwd, \"ToDel\")\n",
    "\n",
    "dir_lib = os.path.join(dir_cwd, \"library\")\n",
    "\n",
    "# -----\n",
    "dir_lib_ena = os.path.join(dir_lib, \"enaBrowserTools\", \"python3_1.6\")\n",
    "\n",
    "dir_tmp_ena = os.path.join(dir_tmp, \"1_ena\")\n",
    "dir_tmp_seq = os.path.join(dir_tmp, \"2_seq\")\n",
    "\n",
    "try:\n",
    "    os.makedirs(dir_tmp_ena)\n",
    "except FileExistsError:\n",
    "    # directory already exists\n",
    "    pass\n",
    "\n",
    "try:\n",
    "    os.makedirs(dir_tmp_seq)\n",
    "except FileExistsError:\n",
    "    # directory already exists\n",
    "    pass\n",
    "\n",
    "# -----\n",
    "file_lib_ena = os.path.join(dir_lib_ena, conf_fname_lib_ena)\n",
    "file_seq_zip = os.path.join(dir_tmp_seq, param_fname_seq_zip)\n",
    "\n",
    "# -----\n",
    "genes_list = param_gene_sequences.split(',') if len(param_gene_sequences) else []\n",
    "\n",
    "if (len(genes_list) == 0):\n",
    "    raise RuntimeError(\"At least 1 gene sequence must be provided\")\n",
    "\n",
    "for gene in genes_list:\n",
    "    # Run ENA browser tools\n",
    "    # os.system(f\"python {file_lib_ena} -f fastq -d {dir_tmp_ena} {gene}\")\n",
    "    os.system(f\"python {file_lib_ena} -f fastq -d {dir_tmp_ena} -c {gene}\")\n",
    "    \n",
    "    if len(os.listdir(os.path.join(dir_tmp_ena, gene))) != 2:\n",
    "        raise RuntimeError(f\"Expected 2 files for gene `{gene}`. Found {len(os.listdir(dir_tmp_ena))} @ {dir_tmp_ena}\")\n",
    "\n",
    "file_fastq_list = glob.glob(os.path.join(dir_tmp_ena, \"*\", \"*.fastq.gz\"))\n",
    "\n",
    "with zipfile.ZipFile(file_seq_zip, \"w\") as zipObj:\n",
    "    for file_fastq in file_fastq_list:\n",
    "        zipObj.write(file_fastq, os.path.basename(file_fastq))\n",
    "        print(f\"Compress gene file {file_fastq}\")\n",
    "print(f\"Generated compressed sequencing files {file_fastq}\")\n",
    "\n",
    "if not os.path.isfile(file_seq_zip):\n",
    "    raise FileNotFoundError(str(file_seq_zip) + \" is missing\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94bc6c68-9b7e-4736-a91c-260dba04b77c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:aneris-dna]",
   "language": "python",
   "name": "conda-env-aneris-dna-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
